### PR TITLE
Added a single test function script and fix debug-test.sh to be more robust

### PR DESCRIPTION
**Pull request by mofosyne**
Continuation of https://github.com/ggerganov/llama.cpp/pull/7096 somewhat to address cases where a person may just want to run a single test.

Also discovered that the test binaries expects to be run in the git root context so need to fix debug-test.sh to take that into account.

@josh-ramer you may want to have a quick look a this to see if this would also be useful to you as well.

This is what the output visually looks like. With emphasis on the last three lines for clearer feedback

image

---

(Created this to try and more quickly pin down a CI issue)

### CODE REVIEW

**Comment by mofosyne** [comment]
Actually @josh-ramer when I ran `./scripts/debug-test.sh test 24` with this script I noticed this printout

```
Ran Test #24: test-eval-callback
GDB args: /home/mofosyne/gitextern/llama.cpp/build-ci-debug/bin/eval-callback "--hf-repo" "ggml-org/models" "--hf-file" "tinyllamas/stories260K.gguf" "--model" "stories260K.gguf" "--prompt" "hello" "--seed" "42" "-ngl" "0"
GDB args @: /home/mofosyne/gitextern/llama.cpp/build-ci-debug/bin/eval-callback~/gitextern/llama.cpp ~/gitextern/llama.cpp/build-ci-debug ~/gitextern/llama.cpp ~/gitextern/llama.cpp
```

Am i reading it correctly? I would have thought that `args[@]` would be the same as `gdb_args[test_number]` but with path expanded, but that doesn't seem to be the case.

**Comment by josh-ramer** [comment]
That comment is a little deceptive, what the code really does is find paths with `/../` in them & replace the `path/../path` with `../path`. There is probably a better solution than that out there if you're pretty skilled in bash scripting. Maybe one easy alternative would be so remove the suffix before `/../`.

**Comment by mofosyne** [comment]
Maybe you can suggest a better comment? Either way, see if this script is working well for your flow and if I didn't break your gdb flow, then well should be good to go!

**Comment by josh-ramer** [comment]
I've got a review started, but there are no EC2 spot instance available right now (first time I've had that happen to my workflow) so I haven't run it yet to make sure it works, but as soon as they're available, I plan to run it for you.

**Review by josh-ramer — APPROVED** [review]
LGTM, will run as soon as I have the compute available.

File: scripts/debug-test.sh
Comment (mofosyne) [reply scripts/debug-test.sh]: for test-json-schema-to-grammar: `./scripts/debug-test.sh test 23` and `./scripts/debug-test.sh -g test 23` functions as expected 

However while `./scripts/debug-test.sh test 24` is working as a test runner `./scripts/debug-test.sh -g test 24` as a test debugger is not functioning as expected. I've rewritten a section to printout your logic in more details and this is what I got for that test:

```
Running Test #24: test-eval-callback
single_test_command: /home/mofosyne/gitextern/llama.cpp/build-ci-debug/bin/eval-callback "--hf-repo" "ggml-org/models" "--hf-file" "tinyllamas/stories260K.gguf" "--model" "stories260K.gguf" "--prompt" "hello" "--seed" "42" "-ngl" "0"
Replacing single test command string that may have /../ in them with the relative path, i.e. path/../path -> ../path.
 - '/home/mofosyne/gitextern/llama.cpp/build-ci-debug/bin/eval-callback' --> '/home/mofosyne/gitextern/llama.cpp/build-ci-debug/bin/eval-callback' 
 - '"--hf-repo' --> '"--hf-repo' 
 - '"ggml-org/models"' --> '"ggml-org/models"' 
 - '"--hf-file"' --> '"--hf-file"' 
 - '"tinyllamas/stories260K.gguf"' --> '"tinyllamas/stories260K.gguf"' 
 - '"--model"' --> '"--model"' 
 - '"stories260K.gguf"' --> '"stories260K.gguf"' 
 - '"--prompt"' --> '"--prompt"' 
 - '"hello"' --> '"hello"' 
 - '"--seed"' --> '"--seed"' 
 - '"42"' --> '"42"' 
 - '"-ngl"' --> '"-ngl"' 
 - '"0"' --> '"0"' 
gdb_args[@]: /home/mofosyne/gitextern/llama.cpp/build-ci-debug/bin/eval-callback"--hf-repo"ggml-org/models""--hf-file""tinyllamas/stories260K.gguf""--model""stories260K.gguf""--prompt""hello""--seed""42""-ngl""0"
```

Key thing to note is the truncation of `"` in `'"--hf-repo' --> '"--hf-repo'` so i suspect a bug in your sed logic.

My SED-ful is not great so I'm going to leave it as is for now and see if you can figure what's going on here.

The other alternative is for us to maybe simply do a direct pass through with some comments explaining the limitation and what should be done in the future.

Comment (josh-ramer) [reply scripts/debug-test.sh]: Yeah, it's an edge case in the regex. I'll take a look later tonight when I get some time.

Comment (josh-ramer) [reply scripts/debug-test.sh]: Ok, I got this working without all the regexs, if you just change the debugger conditional if clause to this:
```
if [ "$gdb_mode" = "true" ]; then
    # Execute debugger
    pushd "$repo_root" || exit 1
    eval "gdb --args ${single_test_command}"
    popd > /dev/null || exit 1
else
```

Comment (josh-ramer) [reply scripts/debug-test.sh]: Another thing to note is that I needed to remove `-DLLAMA_FATAL_WARNINGS=O` from the `make` command in order to get the code to compile due to (maybe worth removing that flag for this script):
```
/home/ubuntu/workspace/llama.cpp/common/common.cpp: In function ‘bool llama_download_file(const std::string&, const std::string&)’:
/home/ubuntu/workspace/llama.cpp/common/common.cpp:2070:48: warning: ignoring attributes on template argument ‘int (*)(FILE*)’ [-Wignored-attributes]
 2070 |         std::unique_ptr<FILE, decltype(&fclose)> outfile(fopen(path_temporary.c_str(), "wb"), fclose);
```

Comment (mofosyne) [reply scripts/debug-test.sh]: Okay, made the changes. Ready to merge

**Comment by mofosyne** [comment]
@josh-ramer ready

**Comment by josh-ramer** [comment]
LGTM

**Comment by mofosyne** [comment]
@josh-ramer see if you can press approve as there needs to be one other approval before it can merge in.

**Review by josh-ramer — APPROVED** [review]
LGTM

**Comment by josh-ramer** [comment]
> @josh-ramer see if you can press approve as there needs to be one other approval before it can merge in.

Looks like I don't have write access. I think that means I'd have to be added as a collaborator like you.

**Review by ggerganov — APPROVED** [review]

**Comment by mofosyne** [comment]
@josh-ramer thanks. It's in now, hope it helps!

**Review by Scutua — APPROVED** [review]

**Pull request merged**
